examples/quantum_variation.py

```diff
@@ -0,0 +1,306 @@
+    # the second and the third term: num/ctrl+num
+    # f: coefficient with simulation gates in paper
+    door = []
+    h_door = []
+    f = []
+    for k in range(l):
+        for i in range(N):
+            f.append(-0.5j)
+            door.append([0, i, numdiff(i)])
+        for i in range(N - 1):
+            f.append(-1j)
+            door.append([5, i, i + 1])
+        for i in range(N - 1):
+            f.append(-1j)
+            door.append([3, i, i + 1])
+    for i in range(N):
+        h.append(1)
+        h_door.append([0, i, numdiff(i)])
+    for i in range(N - 1):
+        h.append(J)
+        h_door.append([5, i, i + 1])
+
+    # initial state
+    state = np.zeros(1 << N)
+    state[0] = 1
+
+    # numerical realize H
+    ls = []
+    weight = []
+    H = np.zeros((1 << N, 1 << N)) * 1j
+    for q in range(len(h_door)):
+        if h_door[q][0] == 0:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 1
+        if h_door[q][0] == 1:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 2
+        if h_door[q][0] == 2:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 3
+        if h_door[q][0] == 3:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 1
+            r[h_door[q][2]] = 1
+        if h_door[q][0] == 4:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 2
+            r[h_door[q][2]] = 2
+        if h_door[q][0] == 5:
+            r = [0 for _ in range(N)]
+            r[h_door[q][1]] = 3
+            r[h_door[q][2]] = 3
+        ls.append(r)
+        weight.append(h[q])
+    ls = tc.array_to_tensor(ls)
+    weight = tc.array_to_tensor(weight)
+    H = tc.quantum.PauliStringSum2Dense(ls, weight, numpy=False)
+
+    # variation realize
+    ODE_theta = np.zeros(len(door))
+    for T in range(int(t / dt)):
+        # calculate coefficient in paper
+        A = np.zeros((len(door), len(door)))
+        C = np.zeros(len(door))
+        batch_theta = []
+        batch_is_k = []
+        batch_is_q = []
+        for k in range(len(door)):
+            for q in range(len(door)):
+                is_k = [0 for _ in range(len(door))]
+                is_k[k] = 1
+                is_q = [0 for _ in range(len(door))]
+                is_q[q] = 1
+                if how_variation == 0:
+                    batch_theta.append(np.angle(f[q]) - np.angle(f[k]))
+                else:
+                    batch_theta.append(np.angle(f[q]) - np.angle(f[k]) - math.pi / 2)
+                batch_is_k.append(is_k)
+                batch_is_q.append(is_q)
+        batch_theta = tc.array_to_tensor(batch_theta)
+        batch_is_k = tf.constant(batch_is_k)
+        batch_is_q = tf.constant(batch_is_q)
+        vmap_result = Calculation_A_vmap(batch_theta, batch_is_k, batch_is_q, ODE_theta)
+        for k in range(len(door)):
+            for q in range(len(door)):
+                A[k, q] = abs(f[k] * f[q]) * vmap_result[k * len(door) + q]
```

Comment:
you can try a vectorized way to assign the result from ``vmap_result`` to ``A``, something like A = f_row matrix multiplication with f_vec and then element wise multiply the reshaped (squared) ``vmap_result``